### PR TITLE
[BEAM-2732] Starting refactor of state tracking in Python

### PR DESCRIPTION
**Pull request by pabloem**
The goal for BEAM-2732 is to refactor the context trackers in the Python SDK so that they will all use the same mechanism.
Currently, Metrics, Logging and StateSampler keep their own contexts. BEAM-2732 aims to have all of them rely on the logic in StateSampler to keep their context (this is already the case in Java).

This PR does the following:
* Clean up legacy code in the StateSampler.
* Adds a global per-thread variable for state sampler, so each execution thread will be able to access the current state sampler from this global variable (this will then be used for LoggingContext, and MetricsEnvironment.current_container).
* Gives the Python-only state sampler functionality to track context, so that non-Cythonized runners can rely on the StateSampler to track context.

### CODE REVIEW

**Comment by pabloem** [comment]
r: @robertwb

**Comment by pabloem** [comment]
Python tests passing.

**Comment by pabloem** [comment]
Run Python PostCommit

File: sdks/python/apache_beam/runners/worker/statesampler_slow_test.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+import logging
+import time
+import unittest
+
+from apache_beam.runners.worker import statesampler_slow as statesampler
+from apache_beam.utils.counters import CounterFactory
+from apache_beam.utils.counters import CounterName
+
+
+class StateSamplerSlowTest(unittest.TestCase):
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/worker/statesampler_slow_test.py]: It would be better to run the same test on both implementations rather than have a copy. (Possibly conditionally test those parts that aren't implemented of course.)

File: sdks/python/apache_beam/runners/worker/statesampler_slow.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+StateSamplerInfo = namedtuple(
+    'StateSamplerInfo',
+    ['state_name', 'transition_count', 'time_since_transition'])
+
+
+# Default period for sampling current state of pipeline execution.
+DEFAULT_SAMPLING_PERIOD_MS = 200
+
+
+class StateSampler(object):
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/worker/statesampler_slow.py]: If we're going to provide a full implementation here, let's make a common base class and put anything that can be shared there. (Possibly the default sampling period as well.)

File: sdks/python/apache_beam/runners/worker/statesampler.pyx [hunk 3]
```diff
@@ -115,11 +121,8 @@ cdef class StateSampler(object):
 
   def __init__(self, prefix, counter_factory,
       sampling_period_ms=DEFAULT_SAMPLING_PERIOD_MS):
-
-    # TODO(pabloem): Remove this once all dashed prefixes are removed from
-    # the worker.
-    # We stop using prefixes with included dash.
-    self.prefix = prefix[:-1] if prefix[-1] == '-' else prefix
+    EXECUTION_STATE_SAMPLERS.set_sampler(self)
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/worker/statesampler.pyx]: That the act of creating a StateSampler should not modify global state.

File: sdks/python/apache_beam/runners/worker/statesampler.pyx [hunk 4]
```diff
@@ -71,19 +71,25 @@ cdef inline int64_t get_nsec_time() nogil:
...
-  def __repr__(self):
-    return ('<StateSamplerInfo state: %s time: %dns transitions: %d>'
-            % (self.state_name,
-               self.time_since_transition,
-               self.transition_count))
+  def set_sampler(self, sampler):
+    self._current_sampler = sampler
+
+
+EXECUTION_STATE_SAMPLERS = ExecutionStateSamplers()
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/worker/statesampler.pyx]: I'd really like to avoid more proliferation of global variables. This is somewhat needed for counters (as the user invokes counter operations without access to the underlying state) but we should be able to pass this state around explicitly.

File: sdks/python/apache_beam/runners/worker/statesampler_slow_test.py (hunk 1 above)
Comment (pabloem) [inline sdks/python/apache_beam/runners/worker/statesampler_slow_test.py]: > **robertwb** wrote:
> It would be better to run the same test on both implementations rather than have a copy. (Possibly conditionally test those parts that aren't implemented of course.)


Acknowledged.

File: sdks/python/apache_beam/runners/worker/statesampler_slow.py (hunk 2 above)
Comment (pabloem) [inline sdks/python/apache_beam/runners/worker/statesampler_slow.py]: > **robertwb** wrote:
> If we're going to provide a full implementation here, let's make a common base class and put anything that can be shared there. (Possibly the default sampling period as well.)


I've restructured it into a class hierarchy. Let me know what you think.

File: sdks/python/apache_beam/runners/worker/statesampler.pyx (hunk 3 above)
Comment (pabloem) [inline sdks/python/apache_beam/runners/worker/statesampler.pyx]: > **robertwb** wrote:
> That the act of creating a StateSampler should not modify global state.


Added a `register` function.

File: sdks/python/apache_beam/runners/worker/statesampler.pyx (hunk 4 above)
Comment (pabloem) [inline sdks/python/apache_beam/runners/worker/statesampler.pyx]: > **robertwb** wrote:
> I'd really like to avoid more proliferation of global variables. This is somewhat needed for counters (as the user invokes counter operations without access to the underlying state) but we should be able to pass this state around explicitly.


I'll document that we want to pass a tracker around as much as possible - but the aim of this global variable is to be able remove global state within MetricsEnvironment, and PerThreadLoggingContext, and have this be the canonical global state.

How does that sound?

**Comment by pabloem** [comment]
Run Python PostCommit

**Comment by pabloem** [comment]
I've created a class hierarchy. [Passes Python PostCommit](https://builds.apache.org/job/beam_PostCommit_Python_Verify/3938/), and [PreCommit](https://builds.apache.org/job/beam_PreCommit_Python_MavenInstall/1801/).

File: sdks/python/apache_beam/runners/worker/statesampler_fast.pyx [hunk 5]
```diff
@@ -109,25 +80,18 @@ cdef class StateSampler(object):
   cdef pythread.PyThread_type_lock lock
 
   cdef public int64_t state_transition_count
-  cdef int64_t time_since_transition
+  cdef int64_t _time_since_transition
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/worker/statesampler_fast.pyx]: Why this change? Or should we be changing all the others as well for consistency?

File: sdks/python/apache_beam/runners/worker/statesampler_fast.pyx [hunk 6]
```diff
@@ -190,60 +158,28 @@ cdef class StateSampler(object):
     if self.started and not self.finished:
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/worker/statesampler_fast.pyx]: Is this method also common to both?

File: sdks/python/apache_beam/utils/counters.py [hunk 7]
```diff
@@ -67,7 +67,7 @@ def __new__(cls, name, stage_name=None, step_name=None,
                                            origin, output_index, io_target)
 
   def __str__(self):
-    return '%s' % self._str_internal()
+    return '<CounterName<%s> at %s>' % (self._str_internal(), hex(id(self)))
```
Comment (robertwb) [inline sdks/python/apache_beam/utils/counters.py]: `__str__` defaults to `__repr__`, if you want them to be the same remove this one.

File: sdks/python/apache_beam/runners/worker/statesampler_test.py [hunk 8]
```diff
@@ -22,20 +22,22 @@
...
 
 
 class StateSamplerTest(unittest.TestCase):
 
   def setUp(self):
     try:
-      # pylint: disable=global-variable-not-assigned
-      global statesampler
-      from apache_beam.runners.worker import statesampler
+      # pylint: disable=unused-variable
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/worker/statesampler_test.py]: You could also put this method in a

```
@classmethod
def setUpClass(cls):
    ...
```

File: sdks/python/apache_beam/runners/worker/statesampler.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    self._current_sampler = None
+
+  def current_sampler(self):
+    return self._current_sampler
+
+  def set_sampler(self, sampler):
+    self._current_sampler = sampler
+
+
+EXECUTION_STATE_SAMPLERS = ExecutionStateSamplers()
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/worker/statesampler.py]: Please get rid of this unused global state.

**Comment by robertwb** [comment]
Let's not add the global variable until it's actually used, in which case we can weigh the pros and cons.

File: sdks/python/apache_beam/runners/portability/fn_api_runner_test.py [hunk 10]
```diff
@@ -23,14 +23,14 @@
...
+from apache_beam.runners.worker import statesampler
 from apache_beam.testing.util import assert_that
 from apache_beam.testing.util import equal_to
 from apache_beam.transforms import window
 
-try:
-  from apache_beam.runners.worker.statesampler import DEFAULT_SAMPLING_PERIOD_MS
-except ImportError:
-  DEFAULT_SAMPLING_PERIOD_MS = 0
+DEFAULT_SAMPLING_PERIOD_MS = 0
```
Comment (pabloem) [inline sdks/python/apache_beam/runners/portability/fn_api_runner_test.py]: > **robertwb** wrote:
> Nit: single assignment is easier to reason about. Put this into an else clause.


Done.

File: sdks/python/apache_beam/runners/worker/statesampler_fast.pyx (hunk 6 above)
Comment (pabloem) [inline sdks/python/apache_beam/runners/worker/statesampler_fast.pyx]: > **robertwb** wrote:
> Is this method also common to both?


You're right. Done.

File: sdks/python/apache_beam/utils/counters.py (hunk 7 above)
Comment (pabloem) [inline sdks/python/apache_beam/utils/counters.py]: > **robertwb** wrote:
> `__str__` defaults to `__repr__`, if you want them to be the same remove this one.


Done.

File: sdks/python/apache_beam/runners/worker/statesampler_fast.pyx (hunk 5 above)
Comment (pabloem) [inline sdks/python/apache_beam/runners/worker/statesampler_fast.pyx]: > **robertwb** wrote:
> Why this change? Or should we be changing all the others as well for consistency?


I had thought this was only c-accessible. I've kept it consistent.

File: sdks/python/apache_beam/runners/worker/statesampler.py (hunk 9 above)
Comment (pabloem) [inline sdks/python/apache_beam/runners/worker/statesampler.py]: > **robertwb** wrote:
> Please get rid of this unused global state.


I plan to use the global state to track metrics and logging (and eventually remove the old style metrics context). I can add this now, or on the next PR. What do you think?
Next PR: https://github.com/apache/beam/pull/4387/files#diff-89072ff532dd15a0af899957de4c26f3R152

File: sdks/python/apache_beam/runners/worker/statesampler_fast.pyx [hunk 11]
```diff
@@ -109,25 +80,18 @@ cdef class StateSampler(object):
...
-      sampling_period_ms=DEFAULT_SAMPLING_PERIOD_MS):
-
-    # TODO(pabloem): Remove this once all dashed prefixes are removed from
-    # the worker.
-    # We stop using prefixes with included dash.
-    self.prefix = prefix[:-1] if prefix[-1] == '-' else prefix
-    self.counter_factory = counter_factory
-    self.sampling_period_ms = sampling_period_ms
+  def __init__(self, *args):
+    #TODO(pabloem): Figure out how to pass arguments without errors.
```
Comment (pabloem) [inline sdks/python/apache_beam/runners/worker/statesampler_fast.pyx]: > **robertwb** wrote:
> What errors were you getting?


hmmm this is fixed now.

**Comment by pabloem** [comment]
Run Python PostCommit

File: sdks/python/apache_beam/runners/worker/statesampler_test.py [hunk 12]
```diff
@@ -22,21 +22,24 @@
...
-  def setUp(self):
+  @classmethod
+  def setUpClass(cls):
     try:
-      # pylint: disable=global-variable-not-assigned
-      global statesampler
-      from apache_beam.runners.worker import statesampler
+      # pylint: disable=unused-variable
+      from apache_beam.runners.worker import statesampler_fast
+      cls.slow_sampler = False
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/worker/statesampler_test.py]: Actually, instead of duplicating the logic, perhaps just query state_sampler.FAST_SAMPLER directly below.

**Comment by pabloem** [comment]
Fixed, and done.

**Comment by robertwb** [comment]
Thanks. Non-Java runs are green. Merging.
